lib/fika/type_checker.ex

```diff
@@ -199,6 +199,29 @@ defmodule Fika.TypeChecker do
     end
   end
 
+  # Map
+  def infer_exp(env, {:map, _, [kv | rest_kvs]}) do
+    {:ok, expected_type, env} = infer_map_key_value(env, kv)
+
+    Enum.reduce_while(rest_kvs, {:ok, expected_type, env}, fn kv, {:ok, acc_type, acc_env} ->
```

Comment:
@sreecodeslayer could you add a to-do comment that we need to refactor this when union types are available?
I'm thinking about working on this support when both mine and your PR is merged, in a third one.

Reply:
Ah i see. Sure will add

Reply:
Actually, @emilsoman how are we gonna handle this case? Should we really support union types for map keys/values?

Reply:
>Should we really support union types for map keys/values?

What would be an usecase for this 🤔

Reply:
I was thinking about JSON, but it seems that keys are always strings. However, if we don't support union-types for values, then we'll need another data-type for decoding JSON

Reply:
```elixir
iex(2)> Jason.encode!(%{1 => :a, "2" => :b}) |> Jason.decode!()
%{"1" => "a", "2" => "b"} 
```

Reply:
ah, based on a similar though of mine for JSON, these can have a dedicated Decoder and Encoder with which we can have custom types through structs. cmiiw @emilsoman

Reply:
We should support union types for keys. A map of type Map(:post|:get, String) can have values like {:post => "foo", :get => "bar"}. This is similar to how lists will support union types too (we'll check if each element "fits into" the list's type). We will have these checks:

1. For maps, subtype?(key, map_key_type) && subtype?(value, map_value_type) should be true
2. For lists, subtype?(elem, list_type) should be true

But since we can't implement everything in one go, we have decided to do this in multiple iterations. First was our simple and stupid "string comparisons" for type checking. Next, we'll add better data structures instead of strings, at this stage, we will can still do to_string on these data structures and keep the comparison unchanged. After that, we'll replace the comparison itself with `subtype?`. Would love to hear your thoughts on this approach too!